### ISSUE TITLE
Error with iglob for **/**

### ISSUE THREAD

**Issue by AbdealiJK** [coala/coala#2664]
I was testing some code in coala, and coala produces the path `/a/b/c/**/**`. In such cases, the output from iglob is invalid.

Here's the actual tree:
```
ajk@localhost site-packages]$ tree /home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear
/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear
├── __init__.py
├── PEP8Bear.py
└── __pycache__
    └── __init__.cpython-35.pyc

1 directory, 3 files
```

```
In [80]: list(iglob("/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/**/**"))
Out[80]: 
['/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/PEP8Bear.py',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__pycache__',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__pycache__/__init__.cpython-35.pyc',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__init__.py',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/PEP8Bear.py/',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__pycache__/',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__pycache__/__init__.cpython-35.pyc',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__pycache__/__init__.cpython-35.pyc/',
 '/home/ajk/miniconda3/envs/py35_coala/lib/python3.5/site-packages/PEP8Bear/__init__.py/']
```

CC @fneu @Adrianzatreanu

**Comment by sils1297** [comment]
@hypothesist please check if we want to backport this to 0.8 it seems like a bug that might affect some users

**Comment by sils1297** [comment]
CC @fneu could you look into this?

**Comment by fneu** [comment]
this is nasty

**Comment by fneu** [comment]
@AbdealiJK I have bad news for you

try passing `recursive=True` to the builtin iglob :sob: :tired_face:

**Comment by fneu** [comment]
it all makes so much sense though.

`**` can match "anything", including "nothing".

the glob pattern "/home/user/**/pig.png" matches `/home/user/pig.png` - see how a slash get's lost?

all of this is kinda consistent and mirrors the behavior of cpython

`>>> fnmatch("/usr/lib/python3.5/site-packages/coala_decorators/__pycache__/decorators.cpython-35.pyc/", "/usr/lib/python3.5/site-packages/coala_decorators/**/**/**")
True`

**Comment by fneu** [comment]
generally, I would consider this "malicious input" much rather than a bug, there is two things I *could* do, though.

a) yield stuff "only" two times. currently, this is possible:
```
>>> from coalib.parsing.Globbing import iglob as myglob
>>> from glob import iglob as cpglob
>>> print("\n".join([x for x in myglob("__pycache__/**/**/**/**/**/**/**/**")]))
__pycache__/
__pycache__/conftest.cpython-35.pyc
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
>>> print("\n".join([x for x in cpglob("__pycache__/**/**/**/**/**/**/**/**", recursive=True)]))
__pycache__/
__pycache__/conftest.cpython-35.pyc
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
__pycache__/conftest.cpython-35.pyc/
```
simply putting the `@yield_once` decorator around `iglob` will reduce the number of appearances of each exact value to one, meaning once with slash and once without.

b) I could also check each value for whether or not it is a file or directory, and add or remove the slash appropriately. This might add some overhead, I'm not sure.

**Comment by fneu** [comment]
concerning b):
usually our (and the cpython) glob results don't contain slashes at the end. Not even for directories. I could therefore just strip them.

```
tests/parsing/GlobTestDir/SubDir1
tests/parsing/GlobTestDir/SubDir1/File11.py
tests/parsing/GlobTestDir/SubDir1/File12.py
```
it's also rather impossible to include slashes in a filename. Windows doesn't allow this, linux neither. I accomplished something similar through a file manager but it seemingly uses another unicode character now. Whoever does this - **especially at the end of a filename **- is acting way more malicious than globbing for "**/**" though, so I guess I'll just do that.

**Comment by fneu** [comment]
@sils1297 fix provided with #2754, overall I don't feel the need to provide a backport for 0.8

**Comment by fneu** [comment]
@AbdealiJK A lot of things are clearer to me now.

the basic problem everybody shares, it that to evaluate `path*1/path*2` inside a glob, you need to figure out all the directories `path*1` matches, and then go from there.

Now the problem is, that if you have a directory with a file and you have a glob `directory/**/**`, the first part, `directory/**`, already matches both the directory and the file.

so for the second `**` you are evaluating `directory/**` and `directory/file/**`, both of which match the file. So you get two results, for everything inside the directory, and potentially more if you look for it.

The iterator-style approach of globbing inside python probably prevents sorting and therefore checks for this, although I added a cache to our glob function now so that it guarantees unique results :+1:
